mlrun/artifacts/base.py

```diff
@@ -324,27 +323,93 @@ def base_dict(self):
+        body_hash = None
+        if not target and not self.spec.target_path:
+            if not mlrun.mlconf.should_generate_target_path_from_artifact_hash():
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Unable to resolve target path, no target path is defined and "
+                    "mlrun.mlconf.artifacts.generate_target_path_from_artifact_hash is set to false"
+                )
+            body_hash, self.spec.target_path = self.resolve_body_target_hash_path(
+                body, artifact_path
+            )
 
-    def _upload_body(self, body, target=None):
-        if calc_hash:
-            self.metadata.hash = blob_hash(body)
+        self.metadata.hash = self.metadata.hash or body_hash
+        if mlrun.mlconf.should_calculate_artifact_hash():
+            self.metadata.hash = calculate_blob_hash(body)
         self.spec.size = len(body)
+
         store_manager.object(url=target or self.spec.target_path).put(body)
 
-    def _upload_file(self, src, target=None):
-        if calc_hash:
-            self.metadata.hash = calculate_local_file_hash(src)
-        self.spec.size = os.stat(src).st_size
-        store_manager.object(url=target or self.spec.target_path).upload(src)
+    def resolve_body_target_hash_path(
+        self, body: typing.Union[bytes, str], artifact_path: str
+    ) -> (str, str):
+        if not artifact_path:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "Unable to resolve body target hash path, artifact_path is not defined"
+            )
+        body_hash = calculate_blob_hash(body)
+        suffix = self._resolve_suffix()
+        artifact_path = (
+            artifact_path + "/" if not artifact_path.endswith("/") else artifact_path
+        )
+        target_path = f"{artifact_path}{body_hash}{suffix}"
+        return body_hash, target_path
+
+    def _upload_file(
+        self, source_path: str, target_path: str = None, artifact_path: str = None
+    ):
+        file_hash = None
+        if not target_path and not self.spec.target_path:
+            if not mlrun.mlconf.should_generate_target_path_from_artifact_hash():
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Unable to resolve target path, no target path is defined and "
+                    "mlrun.mlconf.artifacts.generate_target_path_from_artifact_hash is set to false"
+                )
+            file_hash, self.spec.target_path = self.resolve_file_target_hash_path(
+                source_path, artifact_path
+            )
+        self.metadata.hash = self.metadata.hash or file_hash
+        if mlrun.mlconf.should_calculate_artifact_hash():
+            self.metadata.hash = calculate_local_file_hash(source_path)
+        self.spec.size = os.stat(source_path).st_size
+
+        store_manager.object(url=target_path or self.spec.target_path).upload(
+            source_path
+        )
```

Comment:
This `_upload_file` method is mostly identical to `_upload_body` except the hashing function, the size function and the `upload` instead of `put`.
I suggest creating an `_upload` method which gets the hashing function and size function (actual function objects) and the "upload" function (string which you can then `getattr` from the store_manager). and then `_upload_file` and `_upload_body` can use it without code dup.
That way if any modifications need to be done in the future, they'll only be done once

Reply:
talked offline, merged logic for `resolve_body_target_hash_path` and `resolve_file_target_hash_path` to `_resolve_target_hash_path`. 
Decided to leave `_upload_file` and `upload_body` as moving it to a shared logic function will require passing 4 modification functions which will decrease readability of the functions IMO.